File: ios/Classes/pangle_global/InterstitialAdManager.swift
import Foundation
import PAGAdSDK
import Flutter

class InterstitialAdManager: NSObject {
    private let slotId: String
    private var eventSink: FlutterEventSink?
    private var interstitialAd: PAGLInterstitialAd?
    
    init(slotId: String, eventSink: FlutterEventSink?) {
        self.slotId = slotId
        self.eventSink = eventSink
        super.init()
    }
    
    func loadAd() {
        let request = PAGInterstitialRequest()
        
        PAGLInterstitialAd.load(withSlotID: slotId, request: request) { [weak self] ad, error in
            guard let self = self else { return }
            
            if let error = error {
                DispatchQueue.main.async {
                    let errorCode = (error as NSError).code
                    self.eventSink?(["event": "onAdLoadFailed", "error": ["code": errorCode, "message": error.localizedDescription]])
                }
                return
            }
            
            if let ad = ad {
                self.interstitialAd = ad
                ad.delegate = self
                DispatchQueue.main.async {
                    self.eventSink?(["event": "onAdLoaded"])
                }
            }
        }
    }
    
    func showAd() {
        guard let rootViewController = UIApplication.shared.windows.first?.rootViewController else {
            return
        }
        interstitialAd?.present(fromRootViewController: rootViewController)
    }
}

extension InterstitialAdManager: PAGLInterstitialAdDelegate {
    func adDidShow(_ ad: PAGAdProtocol) {
        eventSink?(["event": "onAdShowed"])
    }
    
    func adDidClick(_ ad: PAGAdProtocol) {
        eventSink?(["event": "onAdClicked"])
    }
    
    func adDidDismiss(_ ad: PAGAdProtocol) {
        eventSink?(["event": "onAdDismissed"])
    }
}

File: android/src/main/kotlin/com/example/multi_ads/pangle_global/InterstitialAdManager.kt
package com.example.multi_ads.pangle_global

import android.app.Activity
import com.bytedance.sdk.openadsdk.api.interstitial.PAGInterstitialAd
import com.bytedance.sdk.openadsdk.api.interstitial.PAGInterstitialAdLoadListener
import com.bytedance.sdk.openadsdk.api.interstitial.PAGInterstitialAdInteractionListener
import com.bytedance.sdk.openadsdk.api.interstitial.PAGInterstitialRequest
import io.flutter.plugin.common.EventChannel

class InterstitialAdManager(
    private val activity: Activity,
    private val slotId: String,
    private val eventSink: EventChannel.EventSink?
) {
    private var interstitialAd: PAGInterstitialAd? = null

    fun loadAd() {
        val request = PAGInterstitialRequest()
        
        PAGInterstitialAd.loadAd(slotId, request, object : PAGInterstitialAdLoadListener {
            override fun onError(code: Int, message: String?) {
                activity.runOnUiThread {
                    eventSink?.success(mapOf(
                        "event" to "onAdLoadFailed",
                        "error" to mapOf(
                            "code" to code,
                            "message" to (message ?: "Unknown error")
                        )
                    ))
                }
            }

            override fun onAdLoaded(ad: PAGInterstitialAd?) {
                interstitialAd = ad
                activity.runOnUiThread {
                    eventSink?.success(mapOf("event" to "onAdLoaded"))
                }
            }
        })
    }

    fun showAd() {
        interstitialAd?.let { ad ->
            ad.setAdInteractionListener(object : PAGInterstitialAdInteractionListener {
                override fun onAdShowed() {
                    eventSink?.success(mapOf("event" to "onAdShowed"))
                }

                override fun onAdClicked() {
                    eventSink?.success(mapOf("event" to "onAdClicked"))
                }

                override fun onAdDismissed() {
                    eventSink?.success(mapOf("event" to "onAdDismissed"))
                }
            })
            
            ad.show(activity)
        }
    }
}
